### ISSUE TITLE
Toggle Line Comment can produce invalid block comment in JSX and CSS (should use single-line comments instead)

### ISSUE THREAD

**Issue by phoenixeliot** [microsoft/vscode#133843]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version:
- OS Version:

Steps to Reproduce:

1. Start with some JSX code
```jsx
function example() {
  return (
    <div>
      <div></div>
      <div></div>
      <div></div>
      <div></div>
      <div></div>
    </div>
  )
}
```
2. Comment out a multi line block
```jsx
function example() {
  return (
    <div>
      <div></div>
      {/* <div></div>
      <div></div>
      <div></div> */}
      <div></div>
    </div>
  )
}
```
3. Comment out an area intersecting the original area
```jsx
function example() {
  return (
    <div>
      {/* <div></div>
      {/* <div></div>
      <div></div>
      <div></div> */}
      <div></div> */}
    </div>
  )
}
```

Result: The comment is invalid and ends prematurely on the first `*/}` instead of the second.

The problem is similar in HTML:
```html
<div>
  <!-- <div></div>
  <!-- <div></div>
  <div></div>
  <div></div> -->
  <div></div> -->
</div>
```

and CSS:
```css
.foo {
  /* a: 1;
  /* b: 2;
  c: 3;
  d: 4; */
  e: 5; */
}
```

# How we should fix it

When commenting multiple lines, it should always use single-line comments for each line, like it already does in JavaScript:
```js
function example() {
  // console.log();
  // // console.log();
  // // console.log();
  // // console.log();
  // console.log();
}
```

I'm not sure exactly what it should do in JSX since you can still have the same "double opening / double closing" problem on one line. Perhaps just add comments to lines that don't have them, and don't add them if they already have them? And if they all have them, then remove them all.

**Comment by alexdima** [comment]
JSX, CSS and HTML do not have line comment tokens (i.e. the logical equivalent of `//`). They only have block comments (i.e. the logical equivalent of `/* ... */`). The respective block comment tokens are: `{/*`, `*/}` for JSX, `/*`,`*/` for CSS and `<!--`, `-->` for HTML.

The toggle comment command falls back to using the block comment tokens in case a language does not support line comment tokens. But this command is not semantical, i.e. it simply inserts `/*`,`*/` at the selection, it doesn't "analyze" the text in any way to figure out that parts of the selection intersects with another block comment, etc.

This has been reported before, let's track in https://github.com/microsoft/vscode/issues/29004 or https://github.com/microsoft/vscode/issues/32541

**Comment by phoenixeliot** [comment]
I don't think this is a duplicate of those issues. The behavior I want is not the same as those two issues. In particular, if lines 3-5 are commented out, I want to be able to comment out lines 2-8 in one go without the inner comment breaking things.

I wanted to share PyCharm's approach here, which I think works great; it escapes the inner comments so they don't prematurely close the block. It also recognizes these when uncommenting a line and un-escapes those sequences back to being a comment, when before they were a double-comment. I much prefer this behavior.
```jsx
function example() {
    return (
        <div>
            {/*<div></div>*/}
            {/*/!*<div></div>*!/*/}
            {/*/!*<div></div>*!/*/}
            {/*/!*<div></div>*!/*/}
            {/*<div></div>*/}
        </div>
    )
}
```
HTML:
```html
<div>
<!--    <div></div>-->
<!--&lt;!&ndash;    <div></div>&ndash;&gt;-->
<!--&lt;!&ndash;    <div></div>&ndash;&gt;-->
<!--&lt;!&ndash;    <div></div>&ndash;&gt;-->
<!--    <div></div>-->
</div>
```
CSS:
```css
.foo {
  /*a: 1;*/
  /*!*b: 2;*!*/
  /*!*c: 3;*!*/
  /*!*d: 4;*!*/
  /*e: 5;*/
}
```

**Comment by alexdima** [comment]
I understand now, thank you for clarifying.
